### DISCUSSION TITLE
Help: A question about "load balance: least_conn" in apisix

### DISCUSSION THREAD

**Discussion by acx-1** [apache/apisix#5548] [Q&A] [answered]
求助，我不太确认目前的least_conn是否存在问题？
       我使用的apisix版本是1.5。打算把最新版本的负载均衡算法（least conn）迁移到项目中。对当前版本的least conn算法存在一点疑问，希望可以得到解答。现在是我发现问题的过程
    1.参考测试用例进行验证的过程中发现，负载均衡的流量统计和预期不同，用三个java服务作为上游，上游中有一台机器的请求加了延迟时间。预期结果：延迟较大的机器实际请求量会少于低延迟的机器；实际结果：三台机器的请求量完全一致，无明显差异；
    2.为了进一步验证，我对比了apisix的[least_conn.lua](https://github.com/apache/apisix/blob/master/apisix/balancer/least_conn.lua)和网上找的[nginx least_conn算法](https://my.oschina.net/u/4313887/blog/3710791) ，发现两边做上游选择的判断公式不同，apisix: （累加1/权重）nginx: (活跃连接数/权值)，
    我不太确认目前的least_conn是否存在问题，希望可以得到解答。

**Comment by spacewander** [comment, marked as answer]
Can you reproduce it with the latest APISIX (2.10.2)?

**Reply by spacewander** [reply]
BTW, our least_conn algorithm is the same as the one used in Kong.

**Reply by acx-1** [reply]
ok,i'll try it later. Actually,i just wana some  test data to comfire this algorithm from somebody using the latest version  becasue i'm using an old version one.

**Reply by acx-1** [reply]
hello，i have read [the latest least_conn algorithm in kong](https://github.com/Kong/kong/blob/master/kong/runloop/balancer/least_connections.lua). And i find something in line 24-45. i think it about  upstream's connections.Is it essential part in this algorithm?could you tell me where can i find this part in apisix?  what the code i saw is Show as below:
```
local function insertAddr(bh, addr)
  addr.connectionCount = addr.connectionCount or 0

  if addr.available then
    bh:insert((addr.connectionCount + 1) / addr.weight, addr)
  end
end

-- @param delta number (+1 or -1) to update the connection count
local function updateConnectionCount(bh, addr, delta)
  addr.connectionCount = addr.connectionCount + delta

  if not addr.available or not bh then
    return
  end

  -- NOTE: we use `connectionCount + 1` this ensures that even on a balancer
  -- with 0 connections the heighest weighted entry is picked first. If we'd
  -- not add the `+1` then any target with 0 connections would always be the
  -- first to be picked (even if it has a very low eight)
  bh:update(addr, (addr.connectionCount + 1) / addr.weight)
end
```

**Reply by spacewander** [reply]
We use `score` in the implementation. `score` is equal to `(addr.connectionCount + 1) / addr.weight`

**Reply by acx-1** [reply]
I try apisix(2.9), and got the expected result. Maybe i am missing something.Thx for answering.

**Comment by acx-1** [comment]
I fixed this bug in my project and shared my solution here.

Scenario: 
    for some reason, I cannot use the latest version. The version I'm using is 1.5. I hope to integrate the latest load balancing algorithm(least_conn) into the project.

Solution: 
    first, I integrated  submissions([#3304](https://github.com/apache/apisix/pull/3304/files)). but i still can't got the expected result. the sum of request each node is always the same.
    After checking, I found that I missed some code  in ([#2001](https://github.com/apache/apisix/pull/2001/files)), which is about after the load balancing algorithm_ The balance method is added to the nginx lifecycle.
